### PR TITLE
[FSTORE-1090] Concepts & Guides for helper columns and on-demand features

### PR DESCRIPTION
**Pull request by davitbzh**

### CODE REVIEW

File: docs/user_guides/fs/feature_view/helper-columns.md [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+using credit card expiry date `expiry_date` that is stored in credit card profile feature group and needs to be fetched from the feature store, as well as date of transaction `date_of_transaction` that arrives at request time. To compute `loc_delta_t_minus_1`
+previous transaction coordinates `longitude` and `latitude` needs to fetched from the feature store and compared to new transaction coordinates that arrives at inference application. In this use case `expiry_date`, `date_of_transaction`, `longitude` and `latitude`
+are `inference_helper_columns`. They are not used for training but are necessary for computing on-demand/derived features.
+
+`training_helper_columns` are a list of feature names that are not the part of the model schema itself but are used during training for the extra information. For example one might want to use feature like `category` of the purchased 
+product to assign different weights during the training time.
+
+## Definition
+Both inference and training helper column name(s) must be part of the `Query` object. If helper column name(s) belong to feature group that is part of a `Join` with `prefix` defined, then this prefix needs to prepended
```
Comment (SirOibaf) [inline docs/user_guides/fs/feature_view/helper-columns.md:15]: Use a `!!! note` section to call out this prefix thing.

File: docs/user_guides/fs/feature_view/helper-columns.md [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+=== "Python"
+
+    !!! example "Define inference and training helper columns with feature views."
+        ```python
+        # define query object 
+        query = label_fg.select("fraud_label")\
+                        .join(cc_profile.select("expiry_date"))\
+                        .join(trans_fg.select(["category", "amount",  "days_until_card_expires", "date_of_transaction"
+                                               "locaton_delta", "longitude", "latitude", "category"])) \
+                        .join(window_aggs_fg.select_except(["trans_volume_mstd", "trans_volume_mavg", "trans_freq",
```
Comment (SirOibaf) [inline docs/user_guides/fs/feature_view/helper-columns.md:27]: Do we need to have some many feature groups here? It's not htat people can copy paste it directly, so I think we should limit the setup code to the bare minimum and focus on the important part, the feature view definition.

File: docs/user_guides/fs/feature_view/helper-columns.md [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+            query=query,
+            labels=["fraud_label"],
+            transformation_functions=transformation_functions,
+            inference_helper_columns=["longitude", "latitude", "date_of_transaction", "expiry_date" ],
+            training_helper_columns=["category"]
+        )
+        ```
+
+## Retrieval
+When replaying a `Query` during model inference, helper columns will be omitted. However, they can be optionally fetched with inference or training data.
```
Comment (SirOibaf) [inline docs/user_guides/fs/feature_view/helper-columns.md:43]: You don't "replay a query" - You are retrieving data for model inference.

File: docs/user_guides/fs/feature_view/helper-columns.md [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+        df['latitute_prev'] = df['latitute'].shift(-1)
+
+        # compute location delta
+        df['loc_delta_t_minus_1'] = df.apply(lambda row: location_delta(row['longitude'], 
+                                                                        row['latitute'],
+                                                                        row['longitude_prev'], 
+                                                                        row['latitute_prev']), axis=1)
+
+        # compute time delta
+        df['days_until_card_expires'] = df.apply(lambda row: time_delta(row['date_of_transaction'],
```
Comment (SirOibaf) [inline docs/user_guides/fs/feature_view/helper-columns.md:73]: Same comment as above, have a single example so that what we care about doesn't get lost in all this code.

File: docs/user_guides/fs/feature_view/helper-columns.md [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+                                             inference_helper['longitude'], 
+                                             inference_helper['latitute'])
+
+        # compute time delta 
+        days_until_card_expires = time_delta(date_of_transaction,
+                                             inference_helper['expiry_date'])
+
+        # Now get assembled feature vector for prediction
+        feature_vector = feature_view.get_feature_vector({"cc_num": cc_num}, 
+                                                          passed_features={"loc_delta_t_minus_1": loc_delta_t_minus_1,
```
Comment (SirOibaf) [inline docs/user_guides/fs/feature_view/helper-columns.md:122]: Same number of example comment.

**Review by SirOibaf — CHANGES_REQUESTED** [review]
The PR also changed the behaviour around primary keys and event time. We should update the Training Data/Batch Data and Feature Vectors page to document this (e.g. how to include the pk or event time int he data returned when getting training data)

File: docs/user_guides/fs/feature_view/batch-data.md [hunk 6]
```diff
@@ -16,6 +16,26 @@ It is very common that ML models are deployed in a "batch" setting where ML pipe
     Dataset<Row> ds = featureView.getBatchData("20220620", "20220627")
     ```
 
+## Retrieve batch data with primary keys and event time
+In certain scenarios, for example for time series model input data needs to be sorted according to primary key(s) and event time combination.  
+Or one might want to merge predictions back to original input data for postmortem analysis. However,  primary key(s) and event time usually are not included in the feature view query as 
+they are not features used for training. To get them pass following attributes to `get_batch_data` method:  
+`primary_keys=True` and/or `event_time=True`.
```
Comment (SirOibaf) [inline docs/user_guides/fs/feature_view/batch-data.md:23]:
```suggestion
For certain use cases, e.g. time series models, the input data needs to be sorted according to the primary key(s) and event time combination. Or one might want to merge predictions back with the original input data for postmortem analysis.
Primary key(s) and event time are not usually included in the feature view query as they are not features used for training.
To retrieve the primary key(s) and/or event time when retrieving batch data for inference, you need to set the parameters `primary_keys=True` and/or `event_time=True`.
```

File: docs/user_guides/fs/feature_view/batch-data.md [hunk 7]
```diff
@@ -16,6 +16,26 @@ It is very common that ML models are deployed in a "batch" setting where ML pipe
...
+    df = feature_view.get_batch_data(
+    start_time = "20220620",
+    end_time = "20220627",
+    primary_keys=True,
+    event_time=True
+    ) # return a dataframe with primary keys and event time
+    ```
+!!! note
+    If event time column has the same name in feature groups included in feature view query then the event time of the left most feature group in the query will be returned. If they have different names then
+    all of them will be returned. Join prefix doesn't have any influence on this behaviour.
```
Comment (SirOibaf) [inline docs/user_guides/fs/feature_view/batch-data.md:37]:
```suggestion
    If the event time columns have the same name across all the feature groups included in the feature view, then only the event time of the label feature group (left most feature group in the query) will be returned. If they have different names, then all of them will be returned. The Join prefix does not have any influence on this behaviour.
```

Tbh I struggle to find the rationale for this behaviour. Why would it matter if the names of the event times are the same or different? Even if the column names are the same, the times might be different, so we either show all of them or only the label feature group. But I don't see the correlation between the names and deciding which column to show.

File: docs/user_guides/fs/feature_view/training-data.md [hunk 8]
```diff
@@ -94,6 +94,25 @@ X_train, X_test, y_train, y_test = feature_view.get_train_test_split(training_da
 X_train, X_val, X_test, y_train, y_val, y_test = feature_view.get_train_validation_test_split(training_dataset_version=1)
 ```
 
+## Read training data with primary key(s) and event time
+In certain scenarios, for example for time series analysis training data needs to be sorted according to primary key(s) and event time combination.  
+However, they usually are not included in the feature view query as they are not features used for training. To get them pass following attributes  
+`primary_keys=True` and/or `event_time=True`.
```
Comment (SirOibaf) [inline docs/user_guides/fs/feature_view/training-data.md:100]:
```suggestion
For certain use cases, e.g. time series models, the input data needs to be sorted according to the primary key(s) and event time combination. 
Primary key(s) and event time are not usually included in the feature view query as they are not features used for training.
To retrieve the primary key(s) and/or event time when retrieving training data, you need to set the parameters `primary_keys=True` and/or `event_time=True`.
```

File: docs/user_guides/fs/feature_view/training-data.md [hunk 9]
```diff
@@ -94,6 +94,25 @@ X_train, X_test, y_train, y_test = feature_view.get_train_test_split(training_da
...
+```python
+# get a training dataset
+X_train, X_test, y_train, y_test = feature_view.get_train_test_split(training_dataset_version=1, 
+                                                                     primary_keys=True,
+                                                                     event_time=True)
+```
+
+!!! note
+    If event time column has the same name in feature groups included in the parent feature view query then the event time of the left most feature group in the query will be returned. If they have different names then
+    all of them will be returned. Join prefix doesn't have any influence on this behaviour.
```
Comment (SirOibaf) [inline docs/user_guides/fs/feature_view/training-data.md:112]:
```suggestion
    If the event time columns have the same name across all the feature groups included in the feature view, then only the event time of the label feature group (left most feature group in the query) will be returned. If they have different names, then all of them will be returned. The Join prefix does not have any influence on this behaviour.
```

File: docs/user_guides/fs/feature_view/helper-columns.md [hunk 10]
```diff
@@ -0,0 +1,170 @@
+---
+description: Using Helper columns in Feature View queries for online/batch inference and training dataset.
+---
+
+# Helper columns
+Hopsworks Feature Store provides functionality to define two types of helper columns `inference_helper_columns` and `training_helper_columns` for [feature views](./overview.md).
```
Comment (SirOibaf) [inline docs/user_guides/fs/feature_view/helper-columns.md:6]:
```suggestion
Hopsworks Feature Store provides a functionality to define two types of helper columns `inference_helper_columns` and `training_helper_columns` for [feature views](./overview.md).
```

File: docs/user_guides/fs/feature_view/helper-columns.md [hunk 11]
```diff
@@ -0,0 +1,170 @@
...
+
+!!! note
+    Both inference and training helper column name(s) must be part of the `Query` object. If helper column name(s) belong to feature group that is part of a `Join` with `prefix` defined, then this prefix needs to prepended
+    to the original column name when defining helper column list.
+
+# Inference Helper columns
```
Comment (SirOibaf) [inline docs/user_guides/fs/feature_view/helper-columns.md:12]: This needs to be a h2 title not h1, i.e., put another # at the beginning

File: docs/user_guides/fs/feature_view/helper-columns.md [hunk 12]
```diff
@@ -0,0 +1,170 @@
...
+            labels=["fraud_label"],
+            transformation_functions=transformation_functions,
+            inference_helper_columns=["longitude", "latitude"],
+        )
+        ```
+
+### Retrieval
+When retrieving data for model inference, helper columns will be omitted. However, they can be optionally fetched with inference or training data.
+
+### Batch inference
```
Comment (SirOibaf) [inline docs/user_guides/fs/feature_view/helper-columns.md:40]: I think you want #### here.

File: docs/user_guides/fs/feature_view/helper-columns.md [hunk 13]
```diff
@@ -0,0 +1,170 @@
...
+        df['loc_delta_t_minus_1'] = df.apply(lambda row: location_delta(row['longitude'], 
+                                                                        row['latitute'],
+                                                                        row['longitude_prev'], 
+                                                                        row['latitute_prev']), axis=1)
+
+        # prepare datatame for prediction
+        df = df[[f.name for f in feature_view.features if not (f.label or f.inference_helper_column or f.training_helper_column)]]
+        ```
+
+### Online inference
```
Comment (SirOibaf) [inline docs/user_guides/fs/feature_view/helper-columns.md:71]: #### here.

**Review by SirOibaf — CHANGES_REQUESTED** [review]

**Review by SirOibaf — APPROVED** [review]
